test/topology_custom/test_tablets_cql.py

```diff
@@ -66,3 +67,65 @@ async def alter_tablets_ks_without_waiting_to_complete():
+    }
+
+    logger.info("starting a node (the leader)")
+    servers = [await manager.server_add(config=config)]
+
+    logger.info("starting a second node (the follower)")
+    servers += [await manager.server_add(config=config)]
+
+    await manager.get_cql().run_async("create keyspace ks with "
+                                      "replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1} and "
+                                      "tablets = {'initial': 2}")
+    await manager.get_cql().run_async("create table ks.t (pk int primary key)")
+
+    logger.info(f"injecting wait-before-committing-rf-change-event into the leader node {servers[0]}")
+    injection_handler = await inject_error_one_shot(manager.api, servers[0].ip_addr,
+                                                    'wait-before-committing-rf-change-event')
+
+    # ALTER tablets KS only accepts a specific DC, it rejects the generic 'replication_factor' tag
+    res = await manager.get_cql().run_async("select data_center from system.local")
+    this_dc = res[0].data_center
+
+    async def alter_tablets_ks_without_waiting_to_complete():
+        logger.info("scheduling ALTER KS to change the RF from 1 to 2")
+        await manager.get_cql().run_async("alter keyspace ks "
+                                          f"with replication = {{'class': 'NetworkTopologyStrategy', '{this_dc}': 2}}")
+
+    # by creating a task this way we ensure it's immediately executed,
+    # but we don't want to wait until the task is completed here,
+    # because we want to do something else in the meantime
+    task = asyncio.create_task(alter_tablets_ks_without_waiting_to_complete())
+
+    logger.info(f"waiting for the leader node {servers[0]} to start handling the keyspace-rf-change request")
+    leader_log_file = await manager.server_open_log(servers[0].server_id)
+    await leader_log_file.wait_for("wait-before-committing-rf-change-event injection hit", timeout=10)
+
+    logger.info(f"creating another keyspace from the follower node {servers[1]} so that the leader, which hangs on injected sleep, "
+                f"wakes up with a changed schema")
+    host = manager.get_cql().cluster.metadata.get_host(servers[1].ip_addr)
+    with disable_schema_agreement_wait(manager.get_cql()):
+        await manager.get_cql().run_async("create keyspace ks2 with "
+                                          "replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1} "
+                                          "and tablets = {'enabled': true}", host=host)
+
+    logger.info("waking up the leader to continue processing ALTER on a changed schema, which should cause a retry")
+    await injection_handler.message()
+
+    logger.info("waiting for ALTER to complete")
+    await task
+
+    # ensure that the concurrent modification error really did take place
+    matches = await leader_log_file.grep("handle_global_request(): concurrent modification, retrying")
```

Comment:
Do you have an alternative to propose?

Reply:
yes, here https://github.com/scylladb/scylladb/pull/21121#discussion_r1808745092

Reply:
> Why this wasn't implemented in test/boost/tablets_test.cc instead?

That's far from a concrete proposal.

 You need a multi-node test to simulate concurrent group 0 commands being sent. It's not possible with a single node.
How do you implement a multi-node test as a boost C++ test? We don't have a framework for that.

Reply:
> you can either simulate the exception

how? perhaps by adding an injection?

> try to add 2 events from the same node

there's a lock that prevents this. Unless we disabled the lock, maybe using another injection (but you complain about there being too many injections)

> whatever works

the test that Piotr wrote works.

Reply:
>the test that Piotr wrote works.

lol, you can't cite my random words out of the context :)

> how? perhaps by adding an injection?

dependency injection is not the same as error injection endpoint, the name is similar indeed

Reply:
This was surely the easiest ways, I'm not sure we want to be hardcore about it.
I also raised a concern if such a test is really needed, I am not sure if we test for conurrent modification in all other places it can happen.
Anyways, additional test is always OK.